File: ballast-core/src/dotnet/Models/AxialCoordinates.cs
using System;

namespace Ballast.Core.Models
{

    public class AxialCoordinatesState
    {
        public int X { get; set; }
        public int Z { get; set; }
    }

    public class AxialCoordinates
    {
        public int X { get; private set; }
        public int Z { get; private set; }

        private AxialCoordinates(int x, int z)
        {
            X = x;
            Z = z;
        }

        public static AxialCoordinates FromProperties(int x, int z) => new AxialCoordinates(
            x: x,
            z: z
        );

        public static AxialCoordinates FromCubic(CubicCoordinates cubic) =>
            new AxialCoordinates(x: cubic.X, z: cubic.Z);

        public static AxialCoordinates FromOffset(OffsetCoordinates offset) =>
            AxialCoordinates.FromCubic(
                CubicCoordinates.FromOffset(offset)
            );

        public static AxialCoordinates FromOrderedPair(int[] orderedPair)
        {
            if (orderedPair.Length < 2)
                throw new Exception("Length of ordered pair must be 2 (or greater)");
            var x = orderedPair[0];
            var z = orderedPair[1];
            return new AxialCoordinates(x, z);
        }

        public bool Equals(AxialCoordinates value)
        {
            if (value == null)
                return false;
            return (
                X == value.X &&
                Z == value.Z
            );
        }

        public bool EqualsCubic(CubicCoordinates cubic) => 
            AxialCoordinates
                .FromCubic(cubic)
                .Equals(this);

        public CubicCoordinates ToCubic() =>  CubicCoordinates.FromAxial(this);

        public OffsetCoordinates ToOffset() => OffsetCoordinates.FromAxial(this);

        public int[] ToOrderedPair() => new int[] { X, Z };

        public static implicit operator AxialCoordinates(AxialCoordinatesState state) =>
            new AxialCoordinates(state.X, state.Z);

    }

}

File: ballast-core/npm/src/domain/models/axial-coordinates.ts
import { CubicCoordinates } from "./cubic-coordinates";
import { OffsetCoordinates } from "./offset-coordinates";

export class AxialCoordinates {

    public readonly x: number;
    public readonly z: number;

    private constructor(x: number, z: number) {
        this.x = x;
        this.z = z;
        if (this.x == -0)
            this.x = 0;
        if (this.z == -0)
            this.z = 0;
    }

    public static fromCubic(object: CubicCoordinates) {
        return new AxialCoordinates(object.x, object.z);
    }

    public static fromOffset(object: OffsetCoordinates) {
        return AxialCoordinates.fromCubic(
            CubicCoordinates.fromOffset(object)
        );
    }

    public static fromOrderedPair(orderedPair: number[]) {
        if (orderedPair.length < 2)
            throw new Error('Length of ordered pair must be 2 (or greater)');
        let x = orderedPair[0];
        let z = orderedPair[1];
        return new AxialCoordinates(x, z);
    }

    public equals(object: AxialCoordinates) {
        if (!object) {
            return false;
        }
        return (
            this.x == object.x &&
            this.z == object.z
        );
    }

    public equalsCubic(object: CubicCoordinates) {
        return AxialCoordinates
            .fromCubic(object)
            .equals(this);
    }

    public toCubic() {
        return CubicCoordinates.fromAxial(this);
    }

    public toOffset() {
        return OffsetCoordinates.fromAxial(this);
    }

    public toOrderedPair() {
        return [this.x, this.z];
    }

}